flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java

```diff
@@ -409,13 +409,30 @@ public static RpcService createRpcService(
 		if (taskManagerHostname != null) {
 			LOG.info("Using configured hostname/address for TaskManager: {}.", taskManagerHostname);
 		} else {
-			Time lookupTimeout = Time.milliseconds(AkkaUtils.getLookupTimeout(configuration).toMillis());
-
-			InetAddress taskManagerAddress = LeaderRetrievalUtils.findConnectingAddress(
-				haServices.getResourceManagerLeaderRetriever(),
-				lookupTimeout);
-
-			taskManagerHostname = taskManagerAddress.getHostName();
+			InetAddress taskManagerAddress;
```

Comment:
I don't think we should get rid of the `LeaderRetrievalUtils.findConnectingAddress` call. The question of whether we want the IP or the hostname should be orthogonal to this.

If the heuristics are not needed, I'm happy to change this in a separate PR.

I would propose something like this:
```java
InetAddress taskManagerAddress = LeaderRetrievalUtils.findConnectingAddress(
  haServices.getResourceManagerLeaderRetriever(),
  lookupTimeout);

String bindPolicy = configuration.getString(TaskManagerOptions.HOST_BIND_POLICY);
switch (bindPolicy.toLowerCase()) {
  case "ip":
    taskManagerHostname = taskManagerAddress.getHostAddress();
    break;
  case "hostname":
    taskManagerHostname = taskManagerAddress.getHostName();
    break;
  default:
    throw new IllegalArgumentException("Unknown " + TaskManagerOptions.HOST_BIND_POLICY.key() + ": " + bindPolicy);
}
```

Reply:
@uce, the problem that with your approach, heuristic is still used and it won't be possible to validate if it's needed or not.
One of the goals of FLINK-11632 is to explicitly separate the heuristic mechanism (in hope that it may be deprecated) and simpler alternative. Extending the heuristic to IP binding would make it harder to get rid of the heuristic mechanism, as it adds additional edge cases (also, there is no request for such feature, afaik).

Reply:
I agree with Ufuk because we might break setups with this change. Either we make configurable whether to take the `ip` or the `hostname` from the heuristically determined `taskManagerAddress` or we make `auto-detect-hostname` the default. 

Additionally we should open a ticket to remove the heuristic/making the default a non-heuristic approach. But before we actually do this, we should discuss it and try to figure out what the consequences are.

Reply:
@tillrohrmann I would actually prefer to remove `auto-detect-hostname` and stick to the proposal above (just `ip` or `hostname`). Having this many options complicates things in my opinion.

@1u0 As another note, why don't we do:
```java
case "ip":
  taskManagerHostName = InetAddress.getByName(taskManagerAddress.getHostName());
  break;
```